fixes #2400 migration to remove orphaned records and add foreign keys constraints to database tables

isratrade opened the pull request.

ohadlevy commented:
[test]

skottler commented:
It looks like the tests failed because of an anomaly. Rerunning them :8ball: 

[test]

isratrade commented:
@skottler, rebased and re-commited

isratrade commented:
rebased after fixed failures in auth_source_ldaps_controller_test

isratrade commented:
@domcleal, I fixed MySQL failures in migration file.
@ohadlevy, I removed FK contracts on reports, logs, messages.  A future ticket can deal with with ON DELETE CASCADE, so we can continue to use .delete_all on Report rather than .destroy_all
I can squash the 2nd commit after tests pass.

Commit:
fixes #2400 migration to remove orphaned records and add foreign keys constraints to database tables

ohadlevy commented:
@domcleal I guess its now or never? any input? :)

isratrade commented:
+1 for referential integrity on the db level

domcleal commented:
I'm for it, but my only concern is with upgrades, and how we can help people who have integrity issues in their DB (whether it's their own fault, or bugs in Foreman, Rails etc).  Have you tried this out on an existing large DB, such as the TLV one?  Maybe we can add some steps to the upgrade notes to help people if the upgrade fails due to the migration.

ohadlevy commented:
@domcleal isnt https://github.com/theforeman/foreman/pull/605/files#L2R4 enough to handle that?

domcleal commented:
Ah, helps if I read the PR, eh?  Yes, that looks like it'd work.

ohadlevy commented on db/migrate/20130808170524_add_keys.rb:
```diff
@@ -0,0 +1,100 @@
+  def change
+
+    # turn off Foreign Key checks in MySQL only.  Postgresl uses "SET CONSTRAINTS ALL DEFERRED;" in test_helper
+    if ActiveRecord::Base.connection.adapter_name == 'Mysql2'
```
what about mysql provider? did we stop using it?

isratrade replied:
does jenkins test for mysql in addition to mysql2?  I added this to get all jenkins tests passing

domcleal replied:
Jenkins only tests with mysql2, but there might be users running with 'mysql', so might be needed there too.

isratrade replied:
Should I change to this 

``` ruby
   if ActiveRecord::Base.connection.adapter_name.downcase =~ /mysql/
```

isratrade replied:
@domcleal, are you going to add a test for adapter 'Mysql' on Jenkins or should I try to do this locally?

domcleal replied:
It'd be better if you can test locally please, I'd rather not add another matrix to the test jobs if I can help it.  I can test this locally too before merge to double check.

isratrade replied:
@domcleal, ok, will do and update

ohadlevy commented:
@isratrade did you validate you can't create an orphan entry via the ui/api? I'm worried that a user will get a wrong error message of db constraint where a simple rails validation / belongs_to/has_many  relationship could fix it?

isratrade commented:
@ohadlevy, can you give an example? In general, orphaned records should only happen if callbacks are not called (delete_all doesn't use :dependant => :destroy whereas destroy_all does).  If a user (which includes API user) goes  through the rails controllers, then all should be fine.  Orphaned records also can happen when a user manipulates the db directory or through rails console and in these cases, the db constraints will help.

isratrade commented:
if using gem 'mysql' rather than gem 'mysql2', when I migrate, here is what I get

```
$ rake db:migrate
"WARNING: Please upgrade to mysql2. The old mysql adapter is not supported by foreigner."
```

isratrade commented:
@domcleal, re-based and will squash after your review

isratrade commented on db/migrate/20130908170524_add_keys.rb:
```diff
@@ -0,0 +1,100 @@
+  def change
+
+    # turn off Foreign Key checks in MySQL only.  Postgresl uses "SET CONSTRAINTS ALL DEFERRED;" in test_helper
+    if ActiveRecord::Base.connection.adapter_name.downcase[0..4] == 'mysql'
```
adapter_name for Mysql is 'MySQL' and for Mysql2 is 'Mysql2'.

domcleal replied:
.start_with? would be much nicer :)

isratrade commented:
rebased and squashed

domcleal commented:
Ah, if the 'mysql' adapter's not supported then that's probably OK, as long as the app still functions.  I'll give it a try in a bit, but first there are some test failures on PostgreSQL and MySQL - could you have a look please?

isratrade commented:
rebased - got a random failure on role_test.  changed from delete_all to destroy_all in role_test

isratrade commented:
rebased - all tests should pass now.  I made some changes to role_test setup.  I could not destroy_all for built-in users.

domcleal commented on db/migrate/20130908100439_delete_orphaned_records.rb:
```diff
@@ -0,0 +1,78 @@
+    KeyPair.where("compute_resource_id NOT IN (?)", ComputeResource.pluck(:id)).destroy_all
+    LookupKey.where("puppetclass_id NOT IN (?)", Puppetclass.pluck(:id)).destroy_all
+    LookupValue.where("lookup_key_id NOT IN (?)", LookupKey.pluck(:id)).destroy_all
+    FactValue.where("fact_name_id NOT IN (?) OR host_id NOT IN (?)", FactName.pluck(:id), Host::Managed.pluck(:id)).destroy_all
```
I think some of these need to reference Host::Base instead of Host::Managed in case users have discovery installed.  In this case I think discovery uploads facts and so there will be facts in the DB referencing a Host::Discovery ID instead.

domcleal commented:
I've packaged foreigner 1.4.2 and immigrant, I'll update the spec at merge time.

domcleal commented:
I merged this prematurely as 4e057da2, so we'll fix the Host::Base issue and spec file in followup commits rather than reverting it.  Thanks @isratrade!
